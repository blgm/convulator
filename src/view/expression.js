import React from 'react'
import { TransitionGroup, CSSTransition } from 'react-transition-group'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Token } from './token'
import { Number } from './number'
import { style } from 'typestyle'

export const expressionStyle = style({
  display: 'flex',
  flexWrap: 'wrap',
  padding: '10px 20px',
  minHeight: '45px'
})

const transitionStyle = style({
  $nest: {
    '&-enter': {
      opacity: 0
    },
    '&-enter-active': {
      opacity: 1,
      transition: 'opacity 250ms ease'
    },
    '&-exit': {
      opacity: 1
    },
    '&-exit-active': {
      opacity: 0,
      transition: 'opacity 250ms ease'
    }
  }
})

export function Expression ({ expression }) {
  return (
    <TransitionGroup className={expressionStyle} >
      {expression.map((token, index) => (
        <CSSTransition
          key={index}
          classNames={transitionStyle}
          timeout={250}
        >
          {token.type === 'number'
            ? <Number value={token.value} />
            : <Token value={token.value} />
          }
        </CSSTransition>
      ))}
    </TransitionGroup>
  )
}
Expression.propTypes = {
  expression: PropTypes.array.isRequired
}

const mapStateToProps = state => ({ expression: state.tokens })

export default connect(mapStateToProps)(Expression)
