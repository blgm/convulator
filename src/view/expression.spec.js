/* eslint-env jest */
import React from 'react'
import {shallow} from 'enzyme'
import {Expression} from './expression'

describe('Expression', () => {
  it('displays nothing when empty', () => {
    const wrapper = shallow(<Expression expression={[]} />)
    const tokens = wrapper.find('TransitionGroup').children()
    expect(tokens).toHaveLength(0)
  })

  it('can display a single number', () => {
    const expression = [4]
    const wrapper = shallow(<Expression expression={expression} />)
    const tokens = wrapper.find('Token')
    expect(tokens).toHaveLength(0)
    const numbers = wrapper.find('Number')
    expect(numbers).toHaveLength(1)
    expect(numbers.at(0).prop('value')).toBe(4)
  })

  it('can display an expression with many tokens', () => {
    const expression = [1, '+', 2, '*', 3]
    const wrapper = shallow(<Expression expression={expression} />)
    const tokens = wrapper.find('TransitionGroup').children()
    expect(tokens.length).toBe(5)
    expect(tokens.at(0).children().at(0).prop('value')).toBe(1)
    expect(tokens.at(1).children().at(0).prop('value')).toBe('+')
    expect(tokens.at(2).children().at(0).prop('value')).toBe(2)
    expect(tokens.at(3).children().at(0).prop('value')).toBe('*')
    expect(tokens.at(4).children().at(0).prop('value')).toBe(3)
  })
})
