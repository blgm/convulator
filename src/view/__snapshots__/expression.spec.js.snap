// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expression can display an expression with many tokens 1`] = `
Array [
  <div
    className="fdwigh6"
  >
    <CSSTransition
      classNames="f1xt770l"
      in={true}
      key=".$0"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Number
          value={4}
        >
          <TransitionGroup
            childFactory={[Function]}
            className="fpwcx99"
            component="div"
          >
            <div
              className="fpwcx99"
            >
              <CSSTransition
                classNames="f16t478l"
                in={true}
                key=".$0"
                onExited={[Function]}
                timeout={250}
              >
                <Transition
                  appear={false}
                  enter={true}
                  exit={true}
                  in={true}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntered={[Function]}
                  onEntering={[Function]}
                  onExit={[Function]}
                  onExited={[Function]}
                  onExiting={[Function]}
                  timeout={250}
                  unmountOnExit={false}
                >
                  <Digit
                    value="4"
                  >
                    <div>
                      4
                    </div>
                  </Digit>
                </Transition>
              </CSSTransition>
            </div>
          </TransitionGroup>
        </Number>
      </Transition>
    </CSSTransition>
    <CSSTransition
      classNames="f1xt770l"
      in={true}
      key=".$1"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Token
          value="+"
        >
          <div
            className="f1lvg2fi"
          >
            +
          </div>
        </Token>
      </Transition>
    </CSSTransition>
    <CSSTransition
      classNames="f1xt770l"
      in={true}
      key=".$2"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Number
          value={2}
        >
          <TransitionGroup
            childFactory={[Function]}
            className="fpwcx99"
            component="div"
          >
            <div
              className="fpwcx99"
            >
              <CSSTransition
                classNames="f16t478l"
                in={true}
                key=".$0"
                onExited={[Function]}
                timeout={250}
              >
                <Transition
                  appear={false}
                  enter={true}
                  exit={true}
                  in={true}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntered={[Function]}
                  onEntering={[Function]}
                  onExit={[Function]}
                  onExited={[Function]}
                  onExiting={[Function]}
                  timeout={250}
                  unmountOnExit={false}
                >
                  <Digit
                    value="2"
                  >
                    <div>
                      2
                    </div>
                  </Digit>
                </Transition>
              </CSSTransition>
            </div>
          </TransitionGroup>
        </Number>
      </Transition>
    </CSSTransition>
    <CSSTransition
      classNames="f1xt770l"
      in={true}
      key=".$3"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Token
          value="*"
        >
          <div
            className="f1lvg2fi"
          >
            x
          </div>
        </Token>
      </Transition>
    </CSSTransition>
    <CSSTransition
      classNames="f1xt770l"
      in={true}
      key=".$4"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Number
          value={3}
        >
          <TransitionGroup
            childFactory={[Function]}
            className="fpwcx99"
            component="div"
          >
            <div
              className="fpwcx99"
            >
              <CSSTransition
                classNames="f16t478l"
                in={true}
                key=".$0"
                onExited={[Function]}
                timeout={250}
              >
                <Transition
                  appear={false}
                  enter={true}
                  exit={true}
                  in={true}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntered={[Function]}
                  onEntering={[Function]}
                  onExit={[Function]}
                  onExited={[Function]}
                  onExiting={[Function]}
                  timeout={250}
                  unmountOnExit={false}
                >
                  <Digit
                    value="3"
                  >
                    <div>
                      3
                    </div>
                  </Digit>
                </Transition>
              </CSSTransition>
            </div>
          </TransitionGroup>
        </Number>
      </Transition>
    </CSSTransition>
  </div>,
  <div
    className="fpwcx99"
  >
    <CSSTransition
      classNames="f16t478l"
      in={true}
      key=".$0"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Digit
          value="4"
        >
          <div>
            4
          </div>
        </Digit>
      </Transition>
    </CSSTransition>
  </div>,
  <div
    className="fpwcx99"
  >
    <CSSTransition
      classNames="f16t478l"
      in={true}
      key=".$0"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Digit
          value="2"
        >
          <div>
            2
          </div>
        </Digit>
      </Transition>
    </CSSTransition>
  </div>,
  <div
    className="fpwcx99"
  >
    <CSSTransition
      classNames="f16t478l"
      in={true}
      key=".$0"
      onExited={[Function]}
      timeout={250}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={250}
        unmountOnExit={false}
      >
        <Digit
          value="3"
        >
          <div>
            3
          </div>
        </Digit>
      </Transition>
    </CSSTransition>
  </div>,
]
`;
